src/main/java/hudson/remoting/Engine.java

```diff
@@ -801,38 +807,115 @@ public void closeRead() throws IOException {
+            justification = "url is provided by the user, and we are trying to connect to it")
+    private Boolean pingSuccessful() throws MalformedURLException {
+        // Unlike JnlpAgentEndpointResolver, we do not use $jenkins/tcpSlaveAgentListener/, as that will be
+        // a 404 if the TCP port is disabled.
+        URL ping = new URL(hudsonUrl, "login");
+        try {
+            HttpURLConnection conn = (HttpURLConnection) ping.openConnection();
+            int status = conn.getResponseCode();
+            conn.disconnect();
+            if (status == 200) {
+                return true;
+            } else {
+                events.status(ping + " is not ready: " + status);
+            }
+        } catch (IOException x) {
+            events.status(ping + " is not ready", x);
+        }
+        return false;
+    }
+
+    private static class ExponentialRetry {
+        final int factor;
+        final Instant beginning;
+        final Duration delay;
+        final Duration timeout;
+        final Duration incrementDelay;
+        final Duration maxDelay;
+
+        ExponentialRetry(Duration timeout) {
+            this(Duration.ofSeconds(0), timeout, 2, Duration.ofSeconds(1), Duration.ofSeconds(10));
+        }
+
+        ExponentialRetry(
+                Duration initialDelay, Duration timeout, int factor, Duration incrementDelay, Duration maxDelay) {
+            this.beginning = Instant.now();
+            this.delay = initialDelay;
+            this.timeout = timeout;
+            this.factor = factor;
+            this.incrementDelay = incrementDelay;
+            this.maxDelay = maxDelay;
+        }
+
+        ExponentialRetry(ExponentialRetry previous) {
+            beginning = previous.beginning;
+            factor = previous.factor;
+            timeout = previous.timeout;
+            incrementDelay = previous.incrementDelay;
+            maxDelay = previous.maxDelay;
+            delay = min(maxDelay, previous.delay.multipliedBy(previous.factor).plus(incrementDelay));
+        }
+
+        private static Duration min(Duration a, Duration b) {
+            return a.compareTo(b) < 0 ? a : b;
+        }
+
+        boolean timeoutExceeded() {
+            return Util.shouldBailOut(beginning, timeout);
+        }
+
+        ExponentialRetry next(EngineListenerSplitter events) throws InterruptedException {
+            var next = new ExponentialRetry(this);
+            if (next.timeoutExceeded()) {
+                events.status("Bailing out after " + DurationFormatter.format(next.timeout));
+                return null;
+            } else {
+                var delaySeconds = next.delay.toSeconds();
+                events.status("Waiting " + delaySeconds + " seconds before retry");
+                TimeUnit.SECONDS.sleep(delaySeconds);
+            }
+            return next;
+        }
+    }
+
+    @FunctionalInterface
+    private interface SupplierThrowingException<T> {
+        T get() throws Exception;
```

Comment:
Why not just `java.util.concurrent.Callable`?

Reply:
True, I often forget about built-in types